lib/src/message_parser.dart

```diff
@@ -0,0 +1,334 @@
+  static List<String> pluralKeywords = [
+    '=0',
+    '=1',
+    '=2',
+    'zero',
+    'one',
+    'two',
+    'few',
+    'many',
+    'other'
+  ];
+  static Map<String, RegExp> pluralKeywordsToRegex = {
+    for (var v in pluralKeywords) v: RegExp('\\s*$v\\s*')
+  };
+  static List<String> genderKeywords = ['female', 'male', 'other'];
+  static Map<String, RegExp> genderKeywordsToRegex = {
+    for (var v in genderKeywords) v: RegExp('\\s*$v\\s*')
+  };
+
+  ///Corresponds to a [+] operator in a regex, matching at least one occurence
+  ///of the [callable].
+  static At<CompositeMessage>? oneOrMore(
+    At? Function(int s) callable,
+    int at,
+  ) {
+    int newAt = -1;
+    List<At> results = [];
+    while (newAt != at) {
+      newAt = at;
+      At? parser = callable(newAt);
+      if (parser != null) {
+        at = parser.at;
+        results.add(parser);
+      }
+    }
+    return results.isNotEmpty
+        ? At(CompositeMessage(results.map((p) => p.result).toList()), newAt)
+        : null;
+  }
+
+  ///Corresponds to an [AND] operator, matching all [callables] or failing, i.e.
+  ///returning [null].
+  static At<CompositeMessage>? and(
+    List<At? Function(int s)> callables,
+    int at,
+  ) {
+    int newAt = at;
+    List<At> resParser = [];
+    for (int i = 0; i < callables.length; i++) {
+      var callable = callables[i];
+      At? parser = callable.call(newAt);
+      if (parser != null) {
+        resParser.add(parser);
+        newAt = parser.at;
+      } else {
+        return null;
+      }
+    }
+    return At(CompositeMessage(resParser.map((p) => p.result).toList()), newAt);
+  }
+
+  ///Match a simple string
+  At<LiteralString>? matchString(int at, String t) => input.startsWith(t, at)
+      ? At(LiteralString(t, null), at + t.length)
+      : null;
+
+  ///Match any of the given keywords
+  At<LiteralString>? asKeywords(Map<String, RegExp> keywordsToRegex, int at) {
+    if (at < input.length) {
+      for (var entry in keywordsToRegex.entries) {
+        var match = entry.value.matchAsPrefix(input, at);
+        if (match != null) {
+          return At(LiteralString(entry.key, null), match.end);
+        }
+      }
+    }
+    return null;
+  }
+
+  ///Parse whitespace
+  At<LiteralString>? trimAt(int at) => at < input.length
+      ? At(LiteralString(input), RegExp(r'\s*').matchAsPrefix(input, at)!.end)
+      : null;
+
+  At<LiteralString>? openCurly(int at) => matchString(at, '{');
+  At<LiteralString>? closeCurly(int at) => matchString(at, '}');
+
+  At<LiteralString>? icuEscapedText(int at) {
+    if (at < input.length) {
+      Match? match = quotedBracketOpen.matchAsPrefix(input, at) ??
+          quotedBracketClose.matchAsPrefix(input, at) ??
+          doubleQuotes.matchAsPrefix(input, at);
+      return match != null
+          ? At(LiteralString(match.group(1)), match.end)
+          : null;
+    }
+    return null;
+  }
+
+  At<LiteralString>? icuText(int at) =>
```

Comment:
I tend to just use expression bodies when the body of the method is just a line or two. You might consider converting most of these multi-line methods to block bodies; I think it would help w/ readability (there's an IntelliJ and VS Code quick fix for switching between the two).

Reply:
I personally prefer to use the expression syntax whenever applicable, as it directly shows me what the return value is. I don't have a very strong preference though :)